Add nextcloud config flow

Hadatko opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds ability to set verify_ssl for Nextcloud integration. (Needed for self-signed certificates)



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #55431
- This PR is related to issue: #55431 #55433
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Hadatko commented:
The code is waiting for testing by author @meichthys

Hadatko commented:
I see that there are some errors not catched by running script on local machine.
But some are not related to my changes i guess:
![image](https://user-images.githubusercontent.com/13130700/197343241-4b3ff98c-b6ea-4a16-a3b0-a6b37cbb09c0.png)

meichthys commented:
I think we may need some changes to the docs to remove the manual config entry setup, and to indicate that it is setup via config flow.

Hadatko commented:
@meichthys were you able to test it?

meichthys commented:
Not yet. Perhaps this evening.

Hadatko commented:
Hi @MartinHjelmare i noticed you made change here. Are you able to configm that build error is not related to our changes?

meichthys commented:
@Hadatko I could have something setup incorrectly, but i'm getting the following when trying to add the nextcloud integration:
<img width="579" alt="Screen Shot 2022-10-27 at 1 15 08 AM" src="https://user-images.githubusercontent.com/10717998/198197903-8acef98d-4c38-4abb-a2f7-be42791f3971.png">
I have your changes pulled in so i'm not sure what's off yet since i don't have time to dig in.

MartinHjelmare commented:
@Hadatko you need to run the scripts mentioned in the build errors and commit the changes.

MartinHjelmare commented on tests/components/nextcloud/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+"""Test the Nextcloud config flow."""
```
We're missing an init module in the test package.

MartinHjelmare commented on homeassistant/components/nextcloud/__init__.py:
```diff
@@ -1,151 +1,28 @@
-SCAN_INTERVAL = timedelta(seconds=60)
-
-# Validate user configuration
-CONFIG_SCHEMA = vol.Schema(
```
We should keep the config schema for a deprecation period while we import the legacy YAML into a config flow with an import flow. See other merged PRs that added config flow for how to use an import step in the config flow and how to create a repairs issue to notify the user.

MartinHjelmare commented on homeassistant/components/nextcloud/const.py:
```diff
@@ -0,0 +1,56 @@
+
+DOMAIN = "nextcloud"
+
+BINARY_SENSORS = (
```
Please move these constants to the respective platforms.

MartinHjelmare commented on homeassistant/components/nextcloud/config_flow.py:
```diff
@@ -0,0 +1,169 @@
+        self.__set_data_points_to_hass()
+
+    def __set_data_points_to_hass(self):
+        self.hass.data[DOMAIN] = self.__get_ncm_data_points()
```
I recommend using our data update coordinator to help fetch the data and keep a data cache.

https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities

Hadatko commented:
> @Hadatko you need to run the scripts mentioned in the build errors and commit the changes.

Did you read the error? If i run script it will remove my dependency. But error is with one different component. Shouldn't you fix that one component instead???

Hadatko commented on homeassistant/components/nextcloud/config_flow.py:
```diff
@@ -0,0 +1,169 @@
+            errors["base"] = "cannot_connect"
+        except UninitializedObject:
+            errors["base"] = "uninitialized"
+        except Exception:  # pylint: disable=broad-except
```
So far i don't think so. Question for @meichthys . Original python module is returning just one exception correct?

meichthys replied:
@Hadatko sorry i just saw this.
Yes, currently nextcloud_monitor only returns a single exeption: NextcloudMonitorError

MartinHjelmare commented:
Try adding the missing `__init__.py` module in the test package and then run `script/gen_requirements_all.py`.

The error is not with a different integration.

Hadatko commented:
> Try adding the missing `__init__.py` module in the test package and then run `script/gen_requirements_all.py`.
> 
> The error is not with a different integration.

Ok then error message is not clear. I see there is warning for other integration not error. I will add missing init script manually then. Thank you for helping with this.

devPeete commented:
Hello, 
@Hadatko are there any news on this PR? I've worked on implementing the config flow for this integration aswell. If you are planning to finish this, we could allign and split the work. I am available on the Discord (Peter#4446).
best regards

Hadatko commented:
@devPeete Can you make PR into my PR in my fork? We can discuss future steps there...

Commit:
Adding config_flow

Not yet working. The Pypi Package does not support aiohttp. Unsure how to work around this.

emontnemery committed:
Bump pychromecast to 13.0.4 (#83896)

Commit:
Merge branch 'fix_ssl_add_config_flow' of https://github.com/devPeete/ha-core into fix_ssl_add_config_flow

epenet committed:
Fix incorrect device class in goodwe (#83917)

Fix device class in goodwe

epenet committed:
Use FREQUENCY device class in integrations (#83916)

* Use FREQUENCY device class in integrations

* Adjust tesla

* Undo goodwe

* Adjust cpuspeed tests

epenet committed:
Refactor huisbaasje to use entity descriptions (#83825)

* Refactor huisbaasje to use entity descriptions

* Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Commit:
Added Async Wrapper for the PyPi Package

cdce8p committed:
Improve ignore_variance typing (#83927)

bachya committed:
Add diagnostics support to PurpleAir (#82446)

Add diagnostics to PurpleAir

epenet committed:
Add stones to mass units (#83933)

bramkragten committed:
Update frontend to 20221213.0 (#83940)

epenet committed:
Refactor Huawei LTE to use Entity Description (#83904)

* Refactor Huawei LTE to use Entity Description

* Apply suggestions from code review

Co-authored-by: Ville Skyttä <ville.skytta@iki.fi>

Co-authored-by: Ville Skyttä <ville.skytta@iki.fi>

epenet committed:
Refactor fitbit descriptions to use device classes (#83936)

bdraco committed:
Bump aioshelly to 5.1.2 to fix state updates not firing after reconnect (#83950)

Bump aioshelly to 5.1.2

changelog: https://github.com/home-assistant-libs/aioshelly/compare/5.1.1...5.1.2

bdraco committed:
Move bluetooth watchdog into the scanner base class (#83888)

synesthesiam committed:
Update intent response (#83858)

* Add language to conversation and intent response

* Move language to intent response instead of speech

* Extend intent response for voice MVP

* Add tests for error conditions in conversation/process

* Move intent response type data into "data" field

* Move intent response error message back to speech

* Remove "success" from intent response

* Add id to target in intent response

* target defaults to None

* Update homeassistant/helpers/intent.py

* Fix test

* Return conversation_id and multiple targets

* Clean up git mess

* Fix linting errors

* Fix more async_handle signatures

* Separate conversation_id and IntentResponse

* Add unknown error code

* Add ConversationResult

* Don't set domain on single entity

* Language is required for intent response

* Add partial_action_done

* Default language in almond agent

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

bdraco committed:
Adjust additional esphome bluetooth debug messages for unique id change (#83955)

bdraco committed:
Bump aioesphomeapi to 13.0.2 to fix reconnects after bad protobuf message (#83951)

bdraco committed:
Try to reconnect disconnected shelly devices as soon as they discovered by zeroconf (#83872)

bdraco committed:
Bump aioshelly to 5.2.0 (#83961)

changelog: https://github.com/home-assistant-libs/aioshelly/compare/5.1.2...5.2.0

synesthesiam committed:
Update intent response (#83962)

* Add language to conversation and intent response

* Move language to intent response instead of speech

* Extend intent response for voice MVP

* Add tests for error conditions in conversation/process

* Move intent response type data into "data" field

* Move intent response error message back to speech

* Remove "success" from intent response

* Add id to target in intent response

* target defaults to None

* Update homeassistant/helpers/intent.py

* Fix test

* Return conversation_id and multiple targets

* Clean up git mess

* Fix linting errors

* Fix more async_handle signatures

* Separate conversation_id and IntentResponse

* Add unknown error code

* Add ConversationResult

* Don't set domain on single entity

* Language is required for intent response

* Add partial_action_done

* Default language in almond agent

* Remove partial_action_done

* Fix linting

* Rename success/fail targets

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

oyvindwe committed:
Fix nobo_hub HVAC mode (#83947)

bachya committed:
Ensure AirVisual Pro uses long-running Samba connection (#83869)

majuss committed:
Bump lupupy dependency to v0.2.3 (#83765)

fixes undefined

devPeete committed:
Added new dependency

ollo69 committed:
Add integration_type property to AsusWrt manifest (#83960)

ollo69 committed:
Use UnitOfTemperature enumerator in AsusWrt (#83958)

engrbm87 committed:
Move transmission to an updated library (#83932)

fixes undefined

epenet committed:
Use UnitOfMass in integrations (#83930)

emontnemery committed:
Ignore certain device trigger validation errors (#83972)

devPeete committed:
reformatting, minor changes, typing

devPeete committed:
Adding checklist generated files

devPeete committed:
Added tests from @Hadatko, did not run them by now

devPeete committed:
Removed Logger Error used for debugging

bachya committed:
Add re-auth flow to PurpleAir (#83445)

* Add re-auth flow to PurpleAir

* Code review

* Code review

* Code review

devPeete commented:
> @devPeete Can you make PR into my PR in my fork? We can discuss future steps there...

I've opened a PR to your repository. Besides testing the config_flow everything is done from the checklist. 
@meichthys do you want to stay Codeowner together with me and @Hadatko?

meichthys commented:
> > @devPeete Can you make PR into my PR in my fork? We can discuss future steps there...
> 
> I've opened a PR to your repository. Besides testing the config_flow everything is done from the checklist. @meichthys do you want to stay Codeowner together with me and @Hadatko?

Things have changed enough that I think you can remove me from being a code owner 👍

Kane610 committed:
Unifi use entity description with sensors (#81930)

* Move bandwidth sensors

* Add uptime sensor

* Use bound

* Fix review comments from other PR

devPeete committed:
Integration is fully typed

devPeete committed:
Removed meichthys as codeowner

Hadatko committed:
Moving consts

Signed-off-by: Dusan Cervenka <cervenka.dusan@gmail.com>

Hadatko committed:
Added test fo nextcloud

Hadatko committed:
Fixing bug from hook

Hadatko committed:
Revert "#80673  Correct fix of ssl for nextcloud"

This reverts commit 6a274e43bb48c81a453a7be7fdd9a93bc76596c4.

Hadatko committed:
#80673  Correct fix of ssl for nextcloud

Hadatko committed:
Added changes which were missing.

Signed-off-by: Dusan Cervenka <cervenka.dusan@gmail.com>

Hadatko committed:
Added application code

Signed-off-by: Dusan Cervenka <cervenka.dusan@gmail.com>

Hadatko committed:
Added missing code owner

Signed-off-by: Dusan Cervenka <cervenka.dusan@gmail.com>

Hadatko committed:
Merge branch 'devPeete-fix_ssl_add_config_flow' into bugfix/nextcloud_ssl_verification

Signed-off-by: Dusan Cervenka <cervenka.dusan@gmail.com>

devPeete commented:
@Hadatko Oh, seems like the Git env of my Linux system was misconfigured. 
I linked the e-maill now to my GitHub account. Hope this is enough.

MartinHjelmare commented:
Please start over in a new PR and branch. The branch is tainted with unrelated commits and you've included 13 other unrelated contributors that will now get notifications for every update of this PR. We're still early in review and won't lose too much history.

Hadatko commented:
@devPeete as you see this PR is closed. I think it would be best if you create new PR
